File: packages/gbnf-experiment/data/20260909T130706Z_31ac2789/ported_implementation/src/grammar_graph/pointers.ts
import type { GraphPointer } from './graph_pointer.ts';

export class Pointers {
  private __pointers__: Map<string, GraphPointer>;

  constructor(...pointers: GraphPointer[]) {
    this.__pointers__ = new Map();
    for (const pointer of pointers) {
      this.add(pointer);
    }
  }

  add(pointer: GraphPointer): void {
    this.__pointers__.set(pointer.id, pointer);
  }

  *[Symbol.iterator](): Generator<GraphPointer> {
    yield* this.__pointers__.values();
  }

  get size(): number {
    return this.__pointers__.size;
  }

  get length(): number {
    return this.size;
  }
}

File: proposed-projects/round-trip-experiment/reverse/20260909T130706Z_31ac2789/20260910T073404Z_10c90cd6/ported_implementation/gbnf/grammar_graph/pointers.py
from typing import Dict, Generator

from .graph_pointer import GraphPointer


class Pointers:
    def __init__(self, *pointers: GraphPointer):
        self.__pointers__: Dict[str, GraphPointer] = {}
        for pointer in pointers:
            self.add(pointer)

    def add(self, pointer: GraphPointer) -> None:
        self.__pointers__[pointer.id] = pointer

    def __iter__(self) -> Generator[GraphPointer, None, None]:
        yield from list(self.__pointers__.values())

    @property
    def size(self) -> int:
        return len(self.__pointers__)

    def __len__(self) -> int:
        return self.size

    def __repr__(self) -> str:
        return f"Pointers({list(self.__pointers__.values())!r})"
